### PR TITLE
Declare strict types

### PR DESCRIPTION
**Pull request by trowski**

### CODE REVIEW

**Review by kelunik — APPROVED** [review]

File: examples/benchmark-memory.php [hunk 1]
```diff
@@ -1,5 +1,7 @@
 <?php
 
+declare(strict_types=1);
```
Comment (Bilge) [inline examples/benchmark-memory.php:3]: Lose the blank line before.

```suggestion
declare(strict_types=1);
```

**Comment by bwoebi** [comment]
Uh, I thought no strict_types were a feature, so that callbacks aren't affected by strict types?

**Comment by kelunik** [comment]
@bwoebi Most people seem to have moved to strict types, no?

**Comment by bwoebi** [comment]
From my experience with proprietary codebases I've seen in the recent years, no.
Many open-source libraries have, that's right.

So, depends on who you want to cater to, obviously.

**Comment by Bilge** [comment]
It's not as if this change prohibits proprietary codebases not using strict types from adoption, unless you can demonstrate otherwise?

**Comment by bwoebi** [comment]
@Bilge Okay, but why do it at all?

**Review by WyriHaximus — APPROVED** [review]

**Comment by kelunik** [comment]
@bwoebi Most proprietary code bases probably contain a lot of legacy code and can't make an instant switch.

**Comment by Crell** [comment]
Strict types is all about making *your* code pickier.  It has no impact on files you don't control, by design.

I am all for making as many things strict as possible, so yes please, make it strict.  Especially as the patch shows there's practically no breakage.

**Comment by trowski** [comment]
> Strict types is all about making your code pickier.

There is an important exception to this: closures. Type strictness on closures is determined by the strictness of the code invoking the callback, not where the closure is defined. As this is a callback-heavy library, our choice here does have consequences to consumers of this library.

In contrast to watcher callbacks in Amp v2's event loop, the callbacks on event watchers here only have a one or two, well-defined parameters. The optional user-data parameter has been eliminated. IMO, I see little impact to invoking event loop callbacks using strict types.

In the time since developing the event loop for Amp v2, it is clear the direction of the open-source PHP community is to use strict types.

**Comment by Bilge** [comment]
@trowski I'm glad someone finally wrote this and it should probably be hoisted to the PR description. But let's be clear: as I mentioned already, strict types do not stop non-strict projects adopting a strictly typed library, even one to which they supply closures. In fact, unless I am grossly mistaken, in the case where the closure specifies no types, there is no impact at all. Only in the case where they do specify types, and then those types are actually mismatched, does there become a slight issue, and one that is easily overcome either with an explicit casting or correctly managing the data in some other way. To be clear, the issue only affects typed closures where the caller was relying on implicit cast semantics. One should expect this case is extremely rare.

**Pull request merged**